ui/src/session.ts

```diff
@@ -74,29 +74,27 @@ interface UpdateSettings extends event.Event<Kind> {
 
 type Msg = ClearCache | Fetch | UpdateSettings;
 
-const fetchSessionRetry = async () => {
-  return api
-    .withRetry(() => api.get<SessionData>(`session`), 200)
-    .then(ses =>
-      sessionStore.success({ status: Status.UnsealedSession, ...ses })
-    )
-    .catch(async error => {
-      if (error instanceof api.ResponseError) {
-        if (error.response.status === 404) {
-          sessionStore.success({ status: Status.NoSession });
-        } else if (error.response.status === 403) {
-          sessionStore.success({ status: Status.SealedSession });
-        } else {
-          throw error;
-        }
-      } else {
-        throw error;
+const fetchSession = async (): Promise<void> => {
+  try {
+    const ses = await api.withRetry(() => api.get<SessionData>(`session`), 200);
+    sessionStore.success({ status: Status.UnsealedSession, ...ses });
+  } catch (err) {
+    if (err instanceof api.ResponseError) {
+      if (err.response.status === 404) {
+        sessionStore.success({ status: Status.NoSession });
+        return;
+      } else if (err.response.status === 403) {
+        sessionStore.success({ status: Status.SealedSession });
+        return;
       }
-    });
-};
+    }
 
-const fetchSession = (): Promise<void> => {
-  return fetchSessionRetry().catch(sessionStore.error);
+    sessionStore.error({
```

Comment:
Is the surfacing of this error to the user part of a follow-up issue?

Reply:
What do you mean by “surfacing”? Setting the error her shows to the blue screen.

Reply:
So I tried to manually break this by pointing the session request to a non-existent endpoint to simulate an error. The app would just go blank and show an error in the console, and when I refresh manually it shows the onboarding start screen.

I expected there to get a notification about the error that the session couldn't be fetched or something like that.

```diff
diff --git a/ui/src/session.ts b/ui/src/session.ts
index 0ed327af..8fb65d0b 100644
--- a/ui/src/session.ts
+++ b/ui/src/session.ts
@@ -76,7 +76,10 @@ type Msg = ClearCache | Fetch | UpdateSettings;

 const fetchSession = async (): Promise<void> => {
   try {
-    const ses = await api.withRetry(() => api.get<SessionData>(`session`), 200);
+    const ses = await api.withRetry(
+      () => api.get<SessionData>(`session-error-please`),
+      200
+    );
     sessionStore.success({ status: Status.UnsealedSession, ...ses });
   } catch (err) {
     if (err instanceof api.ResponseError) {
```

<img width="1508" alt="Screenshot 2020-11-12 at 16 42 13" src="https://user-images.githubusercontent.com/158411/98961658-1bb7e480-2506-11eb-8a25-3b0608350b02.png">

<img width="1552" alt="Screenshot 2020-11-12 at 16 42 27" src="https://user-images.githubusercontent.com/158411/98961670-1f4b6b80-2506-11eb-86d4-9b11f83ecff4.png">

Reply:
The first issue (blank screen with uncaught exception) is due to #1248. It has nothing to do with session errors per se.

The second issue is due to a 404 indicating that the session is missing and that the user needs to be onboarded. The problem is that we’re abusing transport errors (a 404) with application responses (session not available) and we’re unable to distinguish between the two. This is not something that we can address here.